### ISSUE TITLE
Objects with circular dependency break JSON.stringify

### ISSUE THREAD

**Issue by Voyz** [couchbaselabs/node-ottoman#85]
The following code:
```
a = ottoman.model('User', someSchema, someOptions);
b = {ref:a};
a.ref = b;
console.log(a);
```
will produce a `RangeError: Maximum call stack size exceeded` error.

I put together a quick fix for it in modelinstance.js though didn't have time to test it thoroughly. It uses a temp `cache` array to detect circular dependency. Here it is in case it helps:

```
/**
 * Returns a JSON serialized version of this model instance.
 *
 * @returns {Object}
 */
ModelInstance.prototype.toJSON = function() {
	if (!this.loaded()) {
		return null;
	}

	var cloneCache = [];
	function safeClone(obj, depth){
		if (depth > 10000){ // safety check
			console.warn('Depth limit exceeded, returning...');
			return '[Circular depth exceeded]';
		}
		if (cloneCache.indexOf(obj) > -1){ // check for circular dependency
			return '[Circular]';
		}
		var subClone = {};
		cloneCache.push(obj);
		for (var key in obj) {
			if (!obj.hasOwnProperty(key)) continue;
			var value = obj[key];
			if (typeof value === 'object' && value !== null){
				subClone[key] = safeClone(value, depth+1);
			} else {
				subClone[key] = value;
			}
		}
		cloneCache.pop();
		return subClone;
	}

	var clone = safeClone(this, 0);
	return clone;
};

/**
 * A custom inspector to help with debugging of model instances.
 *
 * @private
 * @ignore
 */
var cache = [];
ModelInstance.prototype.inspect = function(depth) {
	if (cache.indexOf(this) > -1){
		return '[Circular]';
	}
	var res = '';
	var name = this.$.schema.name;
	if (!name) {
		name = 'unnamed';
	}
	var attribs = [];
	if (this.$.loaded) {
		attribs.push('loaded');
	} else {
		attribs.push('unloaded');
	}
	attribs.push('key:' + this.$.key);

	res += 'OttomanModel(`' + name + '`, ' + attribs.join(', ') + ', {';
	var hasProperties = false;
	for (var i in this) {
		/* istanbul ignore else */
		if (this.hasOwnProperty(i)) {
			if (!hasProperties) {
				res += '\n';
				hasProperties = true;
			}
			res += '  ' + i + ': ';
			cache.push(this);
			res += util.inspect(this[i]).replace(/\n/g, '\n  ');
			cache.pop();
			res += ',\n';
		}
	}
	res += '})';
	return res;
};
```

**Comment by frank-dspeed** [comment]
this is also a fix for #58

**Comment by brett19** [comment]
This is actually a bit strange.  We do in fact use an internal cache to prevent this issue with circular references.  The only way I can imagine this issue being triggered is if there was an object which was circularly _embedding_ objects, which could never be encoded properly, even with a cache.
